### PR TITLE
Banking

### PR DESCRIPTION
**Pull request by MattGibney**

### CODE REVIEW

File: apps/web-app/src/pages/main/structures/bank/deposit.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    setDepositsRemaining(MAX_DEPOSITS - depositsMade);
+  }, [props.client.authenticatedPlayer]);
+
+  async function handleDeposit(e: React.FormEvent<HTMLFormElement>) {
+    e.preventDefault();
+    const playerGold = props.client.authenticatedPlayer?.gold;
+    if (!playerGold) return;
+
+    if (depositsRemaining < 1) {
+      setInvalidMessages(['You have reached the maximum deposits allowed.']);
```
Comment (fpopa) [inline apps/web-app/src/pages/main/structures/bank/deposit.tsx:37]: this is only validated on the FE right now, which is not good.

File: apps/web-app/src/pages/main/structures/bank/deposit.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      return;
+    }
+
+    const maxDeposit = Math.floor(playerGold * 0.8);
```
Comment (fpopa) [inline apps/web-app/src/pages/main/structures/bank/deposit.tsx:41]: this one as well

File: apps/api/src/models/player.ts [hunk 3]
```diff
@@ -79,6 +91,42 @@ export default class PlayerModel {
     return authedPlayerObject;
   }
 
+  async depositGold(amount: number) {
+    this.gold -= amount;
```
Comment (fpopa) [inline apps/api/src/models/player.ts:95]: we should validate that the amount deposited is <80% of the available gold here.

Also -> what happens If the player is attacked? 🙃 

We should either have a locking mechanism for the gold calculation / attacks / deposits OR achieve the same behaviour by filtering based on the available gold when updating the player available gold:

update player set gold=new_gold, bank_gold=new_bank_gold where player.id=player_id and gold.old_gold.

Wdyt?

File: apps/web-app/src/pages/main/structures/bank/deposit.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    if (playerGoldInBank < 1) {
+      setInvalidMessages(['You do not have any gold in the bank to withdraw.']);
+      return;
+    }
+
+    if (inputAmount > playerGoldInBank) {
+      setInvalidMessages([
+        'You cannot withdraw more gold than you have in the bank',
+      ]);
+      return;
```
Comment (fpopa) [inline apps/web-app/src/pages/main/structures/bank/deposit.tsx:77]: this also needs BE validation.

**Review by fpopa — COMMENTED** [review]
A good step forward 👏 

I commented on a few things that we should definitely fix either in this PR or a subsequent one.

Let me know how you want to progress on this.

Especially the interactions that control the gold need to be done in a clever way in order to avoid race conditions that will mess up calculations.

**Comment by MattGibney** [comment]
I accidentally marked this as ready for review earlier. @fpopa Thanks for the feedback. You're 100% correct about it all

File: apps/api/src/models/player.ts (hunk 3 above)
Comment (MattGibney) [reply apps/api/src/models/player.ts:95]: Validation shouldn't happen here. That'll be done in the controllers.

With regards to locking. I'm not owrried about that at all. If it's something we do want to address, we simply wrap the dao calls in a transaction.

**Comment by MattGibney** [comment]
This is ready for review now. I've opted to not worry about the race condition of a deposit coming in at the same time as an attack. The requests take less than 10ms on the server to process. While it's possible that it someone could attempt to exploit it with a script, I don't think it's a real concern right now. Plus, the correct way to handle that scenario is to use transactions anyway.

**Review by fpopa — APPROVED** [review]
Looks good, let's go 🚀

**Pull request merged**